src/core/connection.cpp

```diff
@@ -33,360 +33,631 @@
 
-    if (!p_websocket_)
-        return;
+        for (auto it = errors.cbegin(); it != errors.cend(); ++it) {
+            const parser::Error &error = *it;
+            std::stringstream error_message;
+            error_message << "parser error: " << error << " \""
+                << Message::to_human_readable(raw_message) << "\"";
+            error_handler_.on_error(error_message.str());
+        }
 
-    state_ = State::DISCONNECTED;
-    p_websocket_->close();
-}
+        if (!errors.empty()) {
+            return;
+        }
 
-State Connection::getConnectionState() { return state_; }
+        parser::MessageList parsed_messages(std::move(parser_result.first));
 
-void Connection::process_messages(Event* p_event, Presence* p_presence)
-{
-    assert(p_event);
-    assert(p_presence);
+        for (auto it = parsed_messages.cbegin(); it != parsed_messages.cend(); ++it) {
+            const Message &parsed_message = *it;
+            DEBUG_MSG("Message received: " << parsed_message.header());
 
-    if (!p_websocket_)
-        return;
+            switch (parsed_message.topic()) {
+                case Topic::EVENT:
+                    event_.notify_(parsed_message);
+                    break;
 
-    Buffer buffer;
-    parser::MessageList messages;
+                case Topic::PRESENCE:
+                    presence_.notify_(parsed_message);
+                    break;
 
-    p_websocket_->set_receive_timeout(std::chrono::milliseconds(1));
+                case Topic::CONNECTION:
+                    handle_connection_response(parsed_message);
+                    break;
 
-    while (receive_(&buffer, &messages) == websockets::State::OPEN) {
-        if (messages.empty())
-            break;
+                case Topic::AUTH:
+                    handle_authentication_response(parsed_message);
+                    break;
 
-        for (const Message& message : messages) {
-            if (message.topic() == Topic::CONNECTION && message.action() == Action::PING) {
-                MessageBuilder pong(Topic::CONNECTION, Action::PONG);
-                send_(pong);
-                continue;
+                default:
+                    {
+                        std::stringstream error_message;
+                        error_message << "unsolicited message: " << parsed_message.header();
+                        error_handler_.on_error(error_message.str());
+                        assert(0);
+                    }
             }
 
-            switch (message.topic()) {
-            case Topic::EVENT:
-                p_event->notify_(message);
-                break;
+        }
+    }
+
+    void Connection::handle_connection_response(const Message &message)
+    {
+        ConnectionState new_state = transition_incoming(state_, message);
 
-            case Topic::PRESENCE:
-                p_presence->notify_(message);
-                break;
+        if (new_state == ConnectionState::ERROR) {
```

Comment:
this part doesn't make much sense. If the client is in an error state, then its a final state. This should probably go into `on_message` since if this is an issue, will be an issue with any type of message.